### ISSUE TITLE
Fix parallel processing on windows

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#718]
After @Ducksual's great work on #713 I believe this is the only remaining bug on Windows.

We're now running CI on appveyor so we do have a test runner: https://ci.appveyor.com/project/drvinceknight/axelrod/build/1.0.17/job/nb5j3y5wrxkho04a

**Comment by Ducksual** [comment]
This might be related to the fact that windows uses 'spawn' instead of 'fork'. I'd be interested to know if some of these tests fail on Linux if you use set_start_method('spawn') somewhere.

https://docs.python.org/3.5/library/multiprocessing.html#contexts-and-start-methods

Going through the current master and removing all of the block multiprocessing on Windows flags gives me 3 test failures. (Possibly more if I've missed some flags).

```
======================================================================
ERROR: test_parallel_play (axelrod.tests.unit.test_tournament.TestTournament)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tests\unit\test_tournament.py", line 259, in test_parallel_play
    scores = tournament.play(processes=2, progress_bar=False).scores
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 116, in play
    self._run_parallel(processes=processes, progress_bar=progress_bar)
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 266, in _start_workers
    process.start()
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: can't pickle generator objects

======================================================================
ERROR: test_progress_bar_play_parallel (axelrod.tests.unit.test_tournament.TestTournament)
Test that tournament plays when asking for progress bar for parallel
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tests\unit\test_tournament.py", line 202, in test_progress_bar_play_parallel
    results = tournament.play(processes=2)
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 116, in play
    self._run_parallel(processes=processes, progress_bar=progress_bar)
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 266, in _start_workers
    process.start()
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'StatusPrinter.<locals>.print_status'

======================================================================
ERROR: test_run_parallel (axelrod.tests.unit.test_tournament.TestTournament)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tests\unit\test_tournament.py", line 286, in test_run_parallel
    self.assertTrue(tournament._run_parallel())
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "C:\Users\ducks\PycharmProjects\Axelrod\axelrod\tournament.py", line 266, in _start_workers
    process.start()
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Miniconda3\envs\axelrod\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'unittest.mock.MagicMock'>: it's not the same object as unittest.mock.MagicMock

----------------------------------------------------------------------
Ran 1747 tests in 121.684s

FAILED (errors=3)
```

**Comment by Ducksual** [comment]
Ok, I had a little look and by inserting a couple of lines at the start of integration/test_tournament.py I can get similar pickling errors, but they're not all the same.

The code:
`from multiprocessing import set_start_method
set_start_method('spawn')`

And the errors:
```
======================================================================
ERROR: test_parallel_play (axelrod.tests.integration.test_tournament.TestTournament)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tests/integration/test_tournament.py", line 62, in test_parallel_play
    scores = tournament.play(processes=2, progress_bar=False).scores
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 116, in play
    self._run_parallel(processes=processes, progress_bar=progress_bar)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 266, in _start_workers
    process.start()
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 274, in _Popen
    return Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 33, in __init__
    super().__init__(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 48, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: can't pickle _csv.writer objects

======================================================================
ERROR: test_parallel_play (axelrod.tests.unit.test_tournament.TestTournament)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tests/unit/test_tournament.py", line 245, in test_parallel_play
    results = tournament.play(processes=2, progress_bar=False)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 116, in play
    self._run_parallel(processes=processes, progress_bar=progress_bar)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 266, in _start_workers
    process.start()
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 274, in _Popen
    return Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 33, in __init__
    super().__init__(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 48, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: can't pickle _csv.writer objects

======================================================================
ERROR: test_progress_bar_play_parallel (axelrod.tests.unit.test_tournament.TestTournament)
Test that tournament plays when asking for progress bar for parallel
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tests/unit/test_tournament.py", line 202, in test_progress_bar_play_parallel
    results = tournament.play(processes=2)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 116, in play
    self._run_parallel(processes=processes, progress_bar=progress_bar)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 266, in _start_workers
    process.start()
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 274, in _Popen
    return Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 33, in __init__
    super().__init__(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 48, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: can't pickle _csv.writer objects

======================================================================
ERROR: test_run_parallel (axelrod.tests.unit.test_tournament.TestTournament)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tests/unit/test_tournament.py", line 286, in test_run_parallel
    self.assertTrue(tournament._run_parallel())
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 230, in _run_parallel
    self._start_workers(workers, work_queue, done_queue)
  File "/Users/ducksual/Documents/PythonProjects/Axelrod-718/axelrod/tournament.py", line 266, in _start_workers
    process.start()
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/context.py", line 274, in _Popen
    return Popen(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 33, in __init__
    super().__init__(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/popen_spawn_posix.py", line 48, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/ducksual/miniconda3/envs/axelrod/lib/python3.5/multiprocessing/reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'unittest.mock.MagicMock'>: it's not the same object as unittest.mock.MagicMock

----------------------------------------------------------------------
Ran 1731 tests in 70.494s

FAILED (errors=5, skipped=1)
```

(I've removed one error about not having matplotlib setup correctly.)

**Comment by drvinceknight** [comment]
Thanks for this @Ducksual.

**Comment by eric-s-s** [comment]
i tried this out on my windows machine using python 3.6 . i only had one issue.  with a little searching i got a workaround (multiprocessing uses pickle, which apparently doesn't mix well with MagicMock.)

```
    def test_run_parallel(self):
        class PickleableMock(MagicMock):
            def __reduce__(self):
                return MagicMock, ()

        tournament = axelrod.Tournament(
            name=self.test_name,
            players=self.players,
            game=self.game,
            turns=axelrod.DEFAULT_TURNS,
            repetitions=self.test_repetitions)
        tournament._write_interactions = PickleableMock(
                    name='_write_interactions')
        self.assertTrue(tournament._run_parallel())

        # Get the calls made to write_interactions
        calls = tournament._write_interactions.call_args_list
        self.assertEqual(len(calls), 15)
```

not sure what to make of that but maybe this is a problem that has solved itself?  

the `PickleableMock` seems to address the error that @DavidCEllis got in `test_run_parallel`

**Comment by marcharper** [comment]
@eric-s-s Want to open a PR?

**Comment by eric-s-s** [comment]
sigh.  not so fast.  on unittest, it works fine, but with nosetest i get (in test_start_workers and test_run_parallel).  this leads me to believe that my workaround was more lucky than right. here's the error in nosetests.

```
  File "C:\Program Files\WinPython-64bit-3.6.1.0Qt5\python-3.6.1.amd64\Lib\multiprocessing\popen_spawn_win32.py", line 65, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Program Files\WinPython-64bit-3.6.1.0Qt5\python-3.6.1.amd64\Lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: can't pickle _thread.RLock objects
```

which means that something is trying to pickle the Queue, maybe (there were a lot of github issues in the google search)?  i'll look further. 

also, does the CI test in a windows environment?

**Comment by eric-s-s** [comment]
someone also wrote a a package called `dill` and a used it to make a multiprocessor in `patheos.multiprocessing` [https://pypi.org/project/pathos/](url) it's supposed to be improved pickling better able to handle curve balls.  i haven't looked at that, but i'll get around to it soon.

**Comment by drvinceknight** [comment]
> which means that something is trying to pickle the Queue, maybe (there were a lot of github issues in the google search)? i'll look further.
> 
> also, does the CI test in a windows environment?

Yes we use appveyor <https://www.appveyor.com/> (which comes up in PRs) to test on windows.

> someone also wrote a a package called dill

I believe we used dill/multiprocess at some point however ended up removing it when we started passing parameters to the parallel workers. I don't know if that means it's not a solution to our problem, just flagging it up.

In case this helps, here's a couple of issues/PRs related to it: 

- https://github.com/Axelrod-Python/Axelrod/issues/518
- https://github.com/Axelrod-Python/Axelrod/pull/522
- https://github.com/Axelrod-Python/Axelrod/pull/529
- https://github.com/Axelrod-Python/Axelrod/issues/563

**Comment by eric-s-s** [comment]
@drvinceknight thanks.  i'll look through stuff and get back to this page if when i find something useful.

**Comment by eric-s-s** [comment]
the issue with noestests seems to be that nosetests runs tests in parallel, which doesn't work with parallel processes.  i'm pushing up a PR that will hopefully work.

**Comment by drvinceknight** [comment]
> the issue with noestests seems to be that nosetests runs tests in parallel, which doesn't work with parallel processes

I'm not too sure I follow. None of our CI test runners use nosetests.

**Comment by drvinceknight** [comment]
(Ping me on gitter if you're around, might be quicker to explain there.)

**Comment by eric-s-s** [comment]
tl;dr  - unix can use os.fork() and doesn't need to pickle everything in Tournament (just some things?) but windows needs to pickle every last bit.  to get multiprocessing working on windows, we would need to isolate/remove the unpickleable bits.  

multiprocessing does work in Windows if `in_memory=True`, `progress_bar=False`, and all the players in `self.players` are pickle-able.  

the three specific areas causing failures: 
- tqdm progress bar
   if `progress_bar=True`, Tournament owns one.  When windows tries to pickle this, it gets ` TypeError:  cannot serialize '_io.TextIOWrapper' object`  ('StatusPrinter.<locals>.print_status' object in @DavidCEllis test results)
- `self.outputfile` and/or `self.writer`. also gets ` TypeError: cannot serialize '_io.TextIOWrapper' object`.
   i tried a quick workaround where Tournament did not keep the file open and only owned self.filename. 
   Every call to `_write_interactions_to_file` used `with open(self.filename, 'w') as f:` and created a new  instance of csv.writer .  I was able to run multiprocessing in windows then and save to file. (i'm not saying this a **good** workaround, just that it didn't fail.)  (note: this was from an extra test i wrote for windows to test parallel play when writing to a specific file location.  will post below.)
- any `SequencePlayer` (i guess just the two ThuMorse) in self.players will cause `TypeError: can't pickle generator objects` .  I don't know if other players are unpickle-able, but if they are, it's the same problem.
   I suppose this could be sidestepped, by storing player classes and init_kwargs instead of instances (i checked, and ThueMorse is pickle-able in this form).

[here's a nice answer about this kind of issue on stackexchange](https://stackoverflow.com/questions/30081961/multiprocessing-works-in-ubuntu-doesnt-in-windows)

This, and the `MagicMock` problem are where the errors are (i think).  @DavidCEllis got all those csv.writer errors on linux because it was trying to pickle a Tournament that owned an opened NamedTemporaryFile and csv.writer (since the default on linux is writing to a file)

**Comment by eric-s-s** [comment]
here was the extra test for writing to a specific file.

```
    def test_parallel_play_with_writing_to_file_on_windows(self):
        """
        The default setting for `play` is to write to a NamedTemporaryFile.
        This is disabled on Windows. This test ensures that parallel_play
        does not conflict with writing the results to file.
        """
        tournament = axelrod.Tournament(
            name=self.test_name,
            players=self.players,
            game=self.game,
            turns=axelrod.DEFAULT_TURNS,
            repetitions=self.test_repetitions)

        results = tournament.play(processes=2, progress_bar=False, filename=self.filename)
        self.assertIsInstance(results, axelrod.ResultSet)
        self.assertEqual(tournament.num_interactions, 75)
```

**Comment by DavidCEllis** [comment]
That TLDR sounds like the conclusion I came to when working on this in the PyCon UK sprints last year. I didn't really get the time to dig into it and figure out exactly what was unpickleable though. The NamedTemporaryFile being the source of the extra issues makes some sense - it's not used in Windows because it doesn't work in windows the way it was being used on unix.

**Comment by drvinceknight** [comment]
> tl;dr - unix can use os.fork() and doesn't need to pickle everything in Tournament (just some things?) but windows needs to pickle every last bit. 

> any SequencePlayer (i guess just the two ThuMorse) in self.players will cause TypeError: can't pickle generator objects . I don't know if other players are unpickle-able, but if they are, it's the same problem.
> I suppose this could be sidestepped, by storing player classes and init_kwargs instead of instances (i checked, and ThueMorse is pickle-able in this form).

Note we already had issues with this which is why the tournament does not actually pass players to the processes but indices of players which refer back to the tournament player list attribute. Obviously something else is still happening.

**Comment by drvinceknight** [comment]
> multiprocessing does work in Windows if in_memory=True, progress_bar=False, and all the players in self.players are pickle-able.

One simple simple "fix" which would end up not disrupting the code base too much would be to overwrite `progress_bar` to `False` and `in_memory` to `True` if on windows and `processes is not None` and display a warning:

```
Warning, you have selected parallel processing which is limited on Windows: no progress bar will be displayed and all calculations will be carried out in memory (which for large tournaments can be very large).
```

I'm not too sure what I think about this but it does make it possible to run parallel processing (as a second citizen) on Windows without disrupting the code base.

Another "solution" (poor use of the word) is to just include a warning that parallel processing is not supported on Windows and leave it at that... I'm also not sure what I think about this.

**Comment by marcharper** [comment]
FYI strategies decorated by a transformer are not pickleable, or were not when we last rewrote this part of the library IIRC. SequencePlayer isn't pickleable because of the generator.

**Comment by eric-s-s** [comment]
@marcharper - interestingly, strategies with strategy_transformers (like BackStabber) are pickle-able, while just making a transformed strategy is not.

**Comment by marcharper** [comment]
Right -- to be pickleable (with the standard library implementation of object serialization) a function or class has to be defined as a top-level object in a module.

**Comment by eric-s-s** [comment]
the only problem that can't be worked around is that all the strategies must be pickle-able. i believe this requires the least amount of change to present code.  I tested all the strategies in axelrod.strategies, and the only problem is `SequencePlayer` . i wrote **a** solution, that works.  here it is.

```
# in class SequencePlayer(Player)
    def __getstate__(self):
        return_dict = self.__dict__.copy()
        del return_dict['sequence_generator']
        return return_dict

    def __setstate__(self, state):
        self.__dict__.update(state)
        self.__dict__['sequence_generator'] = self.generator_function(*self.generator_args)
        for turn in self.history:
            next(self.sequence_generator)
```

and if you added a pickling test to `TestPlayer`, that would cover all the strategies.

```
    def test_player_can_be_pickled(self):
        player = self.player()
        player.play(axl.Cooperator())
        reconstituted = pickle.loads(pickle.dumps(player))
        self.assertEqual(reconstituted, player)
```

It seems that you don't even need the pickled player to be equal, just that their clones are equal (since you are cloning each player before playing a match).  so you could loosen the requirement to be `self.assertEqual(reconstituted.clone(), player.clone())`

>Note we already had issues with this which is why the tournament does not actually pass players to the processes but indices of players which refer back to the tournament player list attribute. 

The above would also be addressed by this.

**Comment by drvinceknight** [comment]
How does this work for users creating their own strategies using transformers? (http://axelrod.readthedocs.io/en/stable/tutorials/advanced/strategy_transformers.html)

The idea being that a user can take any strategy in the library and build a new class with a transformer, those strategies should also be able to be used (in effect there is an infinite number of strategies).

**Comment by drvinceknight** [comment]
(Thanks a lot for working on this @eric-s-s!)

**Comment by drvinceknight** [comment]
> and if you added a pickling test to TestPlayer, that would cover all the strategies.

Perhaps something similar could be added to the strategy transformer tests.

**Comment by eric-s-s** [comment]
riddle me this:
using that same pickling test, i wrote a class like so:

```
import axelrod.strategy_transformers as st

@st.ApologyTransformer([D], [C], name_prefix=None)
class Apology(axl.Cooperator):
    pass
```

when i test `Apology`, it passes.  however, if i remove the `name_prefix=None`, it fails like so:

`_pickle.PicklingError: Can't pickle <class 'axelrod.test_pickling.ApologizingApology'>: attribute lookup ApologizingApology on axelrod.test_pickling failed`

**Comment by marcharper** [comment]
There's also the case of a dynamic strategy creation, where a user at run time e.g. applies a transformer to every strategy (to defect on the first turn) before starting a tournament.

**Comment by drvinceknight** [comment]
Also, fwiw, here's another (parallel processing) test that's skipped on Windows: https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/tests/unit/test_fingerprint.py#L199

**Comment by eric-s-s** [comment]
phew!  i now have a PR that can pickle everything.   however, it will not pass `test_fingerprint.py` . 

In order to get things to work, I had to change how `DualTransformer` was implemented.  The old way made a call to the strategies of the original instance and a new player instance.  The new way only calls the original player instance.  

`AshlockFingerprint` uses a `DualJossAnn`, a stochastic Dual class.  The tests use axl.seed(0).  The results are now different since there are fewer calls to a stochastic method.   

i am confident that the Dual transformer is working correctly, though.  It passed `test_strategy_transformers.py` and the new tests that i wrote.

**Comment by eric-s-s** [comment]
here's the note in `tournament.py` about multiprocessing.

>At first sight, it might seem simpler to use the multiprocessing Pool
   Class rather than Processes and Queues. However, Pool can only accept
   target functions which can be pickled and instance methods cannot.

i did a little checking, and i believe this problem is solved in python3. i checked this with the following code, and it worked just fine in windows.

```
from multiprocessing import Pool
import time
from tqdm import *

class SquareCounts(object):
    def __init__(self, max_num):
        self.max_num = max_num
        self.num_list = range(max_num)

    def reset(self):
        self.num_list = range(self.max_num)

    def get_square(self, num):
        square = num ** 2
        time.sleep(0.5)
        return square

    def writer(self, num, file_obj):
        file_obj.write('{}\n'.format(num))

    def multi(self):
        p = Pool(4)
        with open('delete.txt', 'w') as file:
            for result in tqdm(p.imap(self.get_square, self.num_list),
                               total=self.max_num):
                self.writer(result, file)
        p.close()
        p.join()


if __name__ == '__main__':
    max_ = 50
    obj = SquareCounts(max_)
    obj.multi()
```

would someone mind checking that in linux and see if it works there?

**Comment by drvinceknight** [comment]
> would someone mind checking that in linux and see if it works there?

Worked just fine for me **but** we need to be **very** careful here and make sure the parallel processing doesn't trip up the writing to file (or dict) of the interactions.

This answer/question is mildly relevant:

https://stackoverflow.com/a/37465654

We also need to make sure that the memory consumption is not effected etc... (Just noting down potential pitfalls of changing this very carefully crafted part of the tournament. A lot of blood sweat and tears has been spent on this.)

**Comment by marcharper** [comment]
Also note the Fortran strategies are not (yet) pickable

**Comment by drvinceknight** [comment]
> Also note the Fortran strategies are not (yet) pickable

To give some context to this:

- There is a side library called `axelrod_fortran` https://github.com/Axelrod-Python/axelrod-fortran
- It is a python interface to some original fortran code
- There is an issue open to reverse engineer and subsequently write those strategies in pure python and include them in `axelrod`

TLDR:

There is no need to worry about getting those strategies pickalable. It's very much an edge case so don't worry about it for this issue.

**Comment by drvinceknight** [comment]
> Also note the Fortran strategies are not (yet) pickable

@marcharper and I spoke and his point was that this should not break anything on *nix (whilst I thought he meant that we should ensure the Fortran strategies can be run in parallel on Windows - my mistake).

So... we should make sure that all strategies being pickalable is not a constraint on how parallelisation is done on *nix machines.

**Comment by eric-s-s** [comment]
i tested multiprocess.Join and original flavor.  The current code wins (on a big tournament it's about 50% faster.  If i don't pass a chunksize, it's even worse).  Whatever is done, there should be no conflict with @marcharper 's stipulation.

**Comment by drvinceknight** [comment]
This is now fixed. Thanks @DavidCEllis for the initial work and effort and to @eric-s-s for a the great work hunting it down :clap: :+1:

**Comment by Abhishek-kumar-k** [comment]
AttributeError                            Traceback (most recent call last)
<ipython-input-2-aae2e2683109> in <module>
    288         "patience": 5
    289     })
--> 290     run(args)

<ipython-input-2-aae2e2683109> in run(args)
    212         t_start = time.time()
    213 
--> 214         train_loss, train_auc = train_model(mrnet, train_loader, epoch, num_epochs, optimizer, writer, current_lr, log_every)
    215         val_loss, val_auc = evaluate_model(mrnet, validation_loader, epoch, num_epochs, writer, current_lr)
    216 

<ipython-input-2-aae2e2683109> in train_model(model, train_loader, epoch, num_epochs, optimizer, writer, current_lr, log_every)
     33     losses = []
     34 
---> 35     for i, (image, label, weight) in enumerate(train_loader):
     36         optimizer.zero_grad()
     37 

~\Anaconda3\lib\site-packages\torch\utils\data\dataloader.py in __iter__(self)
    277             return _SingleProcessDataLoaderIter(self)
    278         else:
--> 279             return _MultiProcessingDataLoaderIter(self)
    280 
    281     @property

~\Anaconda3\lib\site-packages\torch\utils\data\dataloader.py in __init__(self, loader)
    717             #     before it starts, and __del__ tries to join but will get:
    718             #     AssertionError: can only join a started process.
--> 719             w.start()
    720             self._index_queues.append(index_queue)
    721             self._workers.append(w)

~\Anaconda3\lib\multiprocessing\process.py in start(self)
    110                'daemonic processes are not allowed to have children'
    111         _cleanup()
--> 112         self._popen = self._Popen(self)
    113         self._sentinel = self._popen.sentinel
    114         # Avoid a refcycle if the target function holds an indirect

~\Anaconda3\lib\multiprocessing\context.py in _Popen(process_obj)
    221     @staticmethod
    222     def _Popen(process_obj):
--> 223         return _default_context.get_context().Process._Popen(process_obj)
    224 
    225 class DefaultContext(BaseContext):

~\Anaconda3\lib\multiprocessing\context.py in _Popen(process_obj)
    320         def _Popen(process_obj):
    321             from .popen_spawn_win32 import Popen
--> 322             return Popen(process_obj)
    323 
    324     class SpawnContext(BaseContext):

~\Anaconda3\lib\multiprocessing\popen_spawn_win32.py in __init__(self, process_obj)
     87             try:
     88                 reduction.dump(prep_data, to_child)
---> 89                 reduction.dump(process_obj, to_child)
     90             finally:
     91                 set_spawning_popen(None)

~\Anaconda3\lib\multiprocessing\reduction.py in dump(obj, file, protocol)
     58 def dump(obj, file, protocol=None):
     59     '''Replacement for pickle.dump() using ForkingPickler.'''
---> 60     ForkingPickler(file, protocol).dump(obj)
     61 
     62 #

AttributeError: Can't pickle local object 'run.<locals>.<lambda>'
